### PR TITLE
Feature/issue 41/write the cli argument parser

### PR DESCRIPTION
**Pull request by frey-perez**
Draft pr for the cli

### CODE REVIEW

**Review by ReinderVosDeWael — COMMENTED** [review]
c.f. https://github.com/childmindresearch/cloai/blob/main/src/cloai/cli/parser.py

Add formatter class "formatter_class": argparse.ArgumentDefaultsHelpFormatter,

File: src/wristpy/core/cli.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def parse_arguments(args: Optional[List[str]] = None) -> argparse.Namespace:
+    """Argument parser for python cli.
+
+    Args:
+        args: Optional argument for when accessed via python script.
+
+    Returns:
+        Namespace object with all of the input arguments.
```
Comment (ReinderVosDeWael) [inline src/wristpy/core/cli.py:20]: It's more than just the inputs right? Default are included.

File: src/wristpy/core/cli.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    parser.add_argument("input", type=pathlib.Path, help="Path to the input data.")
+
+    parser.add_argument(
+        "-o",
+        "--output",
+        type=pathlib.Path,
+        help="Optional path where data will be saved.",
+    )
+
+    parser.add_argument(
```
Comment (ReinderVosDeWael) [inline src/wristpy/core/cli.py:36]: I'm gonna do an annoying teaching moment here: this argument contains a bug. If you resolve the smoke test to a unit test (c.f. comments on that file) you should catch it.

**Review by ReinderVosDeWael — CHANGES_REQUESTED** [review]
I'm missing two important components to this PR:

1. How do I call this from the command line? The standard in Python packages is to use `__main__.py` but that doesn't exist here. This can be resolved by simply moving your `main()` function to a new `__main__.py`. Now the CLI can be called with `python -m wristpy` (c.f. https://docs.python.org/3/library/__main__.html#main-py-in-python-packages). Also add a poetry call (c.f. https://github.com/childmindresearch/cloai/blob/094dd0ac5dd1f314876fe5abbb8c54b0553b8953/pyproject.toml#L47). This lets you run it as `poetry run wristpy` and I _believe_ also as just `wristpy` if you're in the correct environment, though I'd have to verify the latter.
2. We need documentation (i.e. an update to the README) stating how to run this. This need not be extensive (that's what the --help method is for!), but a user should be able to find this. 

On the next review, could you include the output of the `--help` in the PR? I'm like 90% confident that I know what it'll look like from the code, but it'll be helpful to just have it included.

File: src/wristpy/core/cli.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        ],
+        help="Provide three thresholds for light, moderate, and vigorous activity."
+        "values must be given in ascending order. Defaults to config values.",
+    )
+
+    parser.add_argument(
+        "-e",
+        "--epoch_length",
+        type=int,
+        help="Desired sampling rate in seconds. Leave empty to skip downsampling.",
```
[the remark this answers is not in the document]
Comment (frey-perez) [reply src/wristpy/core/cli.py:63]: The epoch length are integers in seconds, so the smallest It can be is 1 second. My assumption is that the user's sampling rate will always be more frequent than 1 Hz. 

To the more general point I took the name from ggir's use of "epoch1", which corresponds to the 5 second temporal resolution, and why we default to 5 second epochs in the orchestrator. 

@Asanto32 Do these two assumptions seem reasonable to you? 

This actually does bring up a question about the default values for this argument, when using the orchestrator in a script the default value is 5, but here when no argument is given it will default to None. Do you consider this inconsistency potentially confusing or is it minor enough to ignore?

File: src/wristpy/core/cli.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        help="Optional path where data will be saved.",
+    )
+
+    parser.add_argument(
+        "-c",
+        "--calibrator",
+        type=str,
+        choices=["ggir", "gradient", "none"],
+        help="Pick which calibrator to use. Can be 'ggir' or 'gradient'."
+        "Leave empty or enter 'none' to proceed without calibration.",
```
[the remark this answers is not in the document]
Comment (frey-perez) [reply src/wristpy/core/cli.py:42]: I actually can't recall why I made it like this, it might've been before I realized that argparse defaults to None. No point in specifying a default if it's just going to be None right?

File: src/wristpy/core/cli.py (hunk 2 above)
Comment (frey-perez) [reply src/wristpy/core/cli.py:36]: Is it related to my comment above ? The line is now: 

```
parser.add_argument(
        "-c",
        "--calibrator",
        type=str,
        choices=["ggir", "gradient"],
        help="Pick which calibrator to use. Can be 'ggir' or 'gradient'.",
    )
```

**Comment by frey-perez** [comment]
> I'm missing two important components to this PR:
> 
> 1. How do I call this from the command line? The standard in Python packages is to use `__main__.py` but that doesn't exist here. This can be resolved by simply moving your `main()` function to a new `__main__.py`. Now the CLI can be called with `python -m wristpy` (c.f. https://docs.python.org/3/library/__main__.html#main-py-in-python-packages). 

Sorry just to clarify, should i move the function definition to __main__.py, or just call the main function there ? My impression was that __main__.py files are usually very minimal as per the document you linked here.

**Comment by ReinderVosDeWael** [comment]
Ah I wasn't aware of that recommendation, yeah you could have `__main__.py` jsut call `cli.main()`

File: src/wristpy/core/cli.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Asanto32) [reply src/wristpy/core/cli.py:62]: Maybe rephrase help to mention that this is the window length for base metrics (enmo, anglez, activity)

**Comment by frey-perez** [comment]
Ready for review again, here's the help command print out for you @ReinderVosDeWael 

```
Run the main wristpy pipeline

positional arguments:
  input                 Path to the input data.

options:
  -h, --help            show this help message and exit
  -o OUTPUT, --output OUTPUT
                        Path where data will be saved. Supports .csv and .parquet formats. (default: None)
  -c {ggir,gradient,none}, --calibrator {ggir,gradient,none}
                        Pick which calibrator to use. Can be 'ggir' or 'gradient'. (default: none)
  -t THRESHOLDS THRESHOLDS THRESHOLDS, --thresholds THRESHOLDS THRESHOLDS THRESHOLDS
                        Provide three thresholds for light, moderate, and vigorous activity.Values must be given in ascending order. (default: [0.03, 0.1, 0.3])
  -e EPOCH_LENGTH, --epoch_length EPOCH_LENGTH
                        Specify the sampling rate in seconds for all metrics. Set to 0 toskip downsampling. (default: 5)
  -v, --version         show program's version number and exit

```

File: README.md [hunk 5]
```diff
@@ -49,43 +49,35 @@ pip install git+https://github.com/childmindresearch/wristpy
...
-
-enmo_epoch1 = computations.moving_mean(enmo)
-anglez_epoch1 = computations.moving_mean(anglez)
-
-#Find sleep windows
-sleep_detector_class = analytics.GGIRSleepDetection(anglez)
-sleep_windows = sleep_detector_class.run_sleep_detection()
+### Using Wristpy through the command-line:
+```sh
+python wristpy /input/file/path.gt3x -o /save/path/file_name.csv -c gradient
```
Comment (ReinderVosDeWael) [inline README.md:54]: It should be possible to run it as just "wristpy" if it's installed in the current environment.

File: src/wristpy/core/cli.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            will take the args from `sys.argv`.
+
+    Returns:
+        Namespace object with all of the input arguments and default values.
+    """
+    default_settings = config.Settings()
+    parser = argparse.ArgumentParser(
+        description="Run the main wristpy pipeline",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        epilog="Please report issues at https://github.com/childmindresearch/wristpy",
```
Comment (ReinderVosDeWael) [inline src/wristpy/core/cli.py:26]: I don't see the epilog in the help copy you posted.

File: src/wristpy/core/cli.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    Args:
+        args: A list of command line arguments given as strings. If None, the parser
+            will take the args from `sys.argv`.
+
+    Returns:
+        Namespace object with all of the input arguments and default values.
+    """
+    default_settings = config.Settings()
+    parser = argparse.ArgumentParser(
+        description="Run the main wristpy pipeline",
```
Comment (ReinderVosDeWael) [inline src/wristpy/core/cli.py:24]: Throughout these help messages: do not forget punctuation ;).

File: src/wristpy/core/cli.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        ],
+        help="Provide three thresholds for light, moderate, and vigorous activity."
+        "Values must be given in ascending order.",
+    )
+
+    parser.add_argument(
+        "-e",
+        "--epoch_length",
+        default=5,
+        type=int,
```
Comment (ReinderVosDeWael) [inline src/wristpy/core/cli.py:64]: This is correct, though @Asanto32 @frey-perez do we want to make an issue out of allowing floats? This would require updates elsewhere so beyond the scope of this PR either way.

File: tests/smoke/test_cli_smoke.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
+"""Smoke tests for wristpy cli."""
+
+import pathlib
+
+import pytest_mock
+
+from wristpy.core import cli, config, orchestrator
+
+
+def test_main_default(
```
Comment (ReinderVosDeWael) [inline tests/smoke/test_cli_smoke.py:10]: This is now, strictly, testing the function as a unit. I.e. this need not be in the smoke directory anymore.

File: tests/unit/test_analytics.py [hunk 10]
```diff
@@ -233,19 +233,6 @@ def test_run_sleep_detection(sleep_detection: analytics.GGIRSleepDetection) -> N
     assert isinstance(result, List)
```
Comment (ReinderVosDeWael) [inline tests/unit/test_analytics.py:235]: Why is this removed?

Comment (frey-perez) [reply tests/unit/test_analytics.py:235]: I believe I've found the commit that made this change, and It's a merge from the orchestrator branch, 

https://github.com/childmindresearch/wristpy/pull/104/commits/7d54381f3baa616d975a7201e59b991182384ba4

File: tests/smoke/test_cli_smoke.py (hunk 9 above)
Comment (frey-perez) [reply tests/smoke/test_cli_smoke.py:10]: Okay, moved main tests to cli unit tests.

File: src/wristpy/core/cli.py (hunk 6 above)
Comment (frey-perez) [reply src/wristpy/core/cli.py:26]: My fault, I cut it off by mistake: 

```
usage: wristpy [-h] [-o OUTPUT] [-c {ggir,gradient,none}] [-t THRESHOLDS THRESHOLDS THRESHOLDS] [-e EPOCH_LENGTH] [-v] input

Run the main wristpy pipeline

positional arguments:
  input                 Path to the input data.

options:
  -h, --help            show this help message and exit
  -o OUTPUT, --output OUTPUT
                        Path where data will be saved. Supports .csv and .parquet formats. (default: None)
  -c {ggir,gradient,none}, --calibrator {ggir,gradient,none}
                        Pick which calibrator to use. Can be 'ggir' or 'gradient'. (default: none)
  -t THRESHOLDS THRESHOLDS THRESHOLDS, --thresholds THRESHOLDS THRESHOLDS THRESHOLDS
                        Provide three thresholds for light, moderate, and vigorous activity.Values must be given in ascending order. (default: [0.03, 0.1, 0.3])
  -e EPOCH_LENGTH, --epoch_length EPOCH_LENGTH
                        Specify the sampling rate in seconds for all metrics. Set to 0 toskip downsampling. (default: 5)
  -v, --version         show program's version number and exit

Please report issues at https://github.com/childmindresearch/wristpy
```

File: tests/unit/test_analytics.py (hunk 10 above)
Comment (frey-perez) [reply tests/unit/test_analytics.py:235]: OH wait, sorry I confused myself, I just remembered the point of this. We added the model validator to the Settings, so we decided to get rid of all these other checks on the thresholds, as that's now covered by the validator.

**Review by ReinderVosDeWael — APPROVED** [review]
